heimctl/retention/scanner.go

```diff
@@ -0,0 +1,80 @@
+package retention
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"euphoria.io/heim/backend/cluster"
+	"euphoria.io/heim/backend/psql"
+	"euphoria.io/scope"
+)
+
+const (
+	maxErrors = 3
+)
+
+func scanForExpired(ctx scope.Context, c cluster.Cluster, pb *psql.Backend) error {
+	rows, err := pb.DbMap.Select(
+		psql.Room{},
+		"SELECT name, founded_by, retention_days FROM room")
+	if err != nil {
+		return err
+	}
+	for _, row := range rows {
+		room, ok := row.(*psql.Room)
+		if !ok {
+			fmt.Printf("error: expected row of type *psql.Room, got %T\n", row)
+			continue
+		}
+		var oldest time.Time
+		err := pb.DbMap.SelectOne(&oldest,
+			"SELECT Min(posted) FROM message WHERE room = $1",
+			room.Name)
+		// no rows => no expired messages
+		if err == sql.ErrNoRows {
+			roomHasExpiredMsg.With(prometheus.Labels{"room": room.Name}).Set(0)
+			continue
+		}
+		if err != nil {
+			fmt.Printf("error selecting oldest message: %s\n", err)
+			continue
+		}
+		// convert RetentionDays to seconds to compare to oldest
+		threshold := time.Now().Add(time.Duration(-room.RetentionDays) * 24 * time.Hour)
```

Comment:
It occurs to me that we need a grace period. In the steady state, messages will be expiring as frequently as they were posted. We shouldn't count them without giving the reaper a chance to reap.

```
const GracePeriod = time.Hour

threshold := time.Now().Add(time.Duration(-room.RetentionDays) * 24 * time.Hour - GracePeriod)
```